bitcoin/src/blockdata/script/witness_program.rs

```diff
@@ -46,11 +50,61 @@ impl WitnessProgram {
         Ok(WitnessProgram { version, program })
     }
 
+    /// Creates a [`WitnessProgram`] from `pk` for a P2WPKH output.
+    pub fn p2wpkh(pk: &PublicKey) -> Result<Self, key::UncompressedPubkeyError> {
+        let hash = pk.wpubkey_hash()?;
+        let program = WitnessProgram::new(WitnessVersion::V0, hash)
+            .expect("wpubkey hash is a valid 20 byte program for segwit v0");
+
+        Ok(program)
+    }
+
+    /// Creates a [`WitnessProgram`] from `script` for a P2WSH output.
+    pub fn p2wsh(script: &Script) -> Self {
+        let hash = script.wscript_hash();
+        WitnessProgram::new(WitnessVersion::V0, hash)
+            .expect("script hash is a valid 20 byte program for segwit v0")
+    }
+
+    /// Creates a pay to taproot address from an untweaked key.
+    pub fn p2tr<C: Verification>(
+        secp: &Secp256k1<C>,
+        internal_key: UntweakedPublicKey,
+        merkle_root: Option<TapNodeHash>,
+    ) -> Self {
+        let (output_key, _parity) = internal_key.tap_tweak(secp, merkle_root);
+        let ser = output_key.to_inner().serialize();
+        WitnessProgram::new(WitnessVersion::V1, ser)
+            .expect("serialized taproot key is a valid program")
+    }
+
+    /// Creates a pay to taproot address from a pre-tweaked output key.
+    ///
+    /// This method is not recommended for use, [`Self::p2tr()`] should be used where possible.
```

Comment:
I think this phrase was copied from somewhere and is no longer true since we statically check that the key is tweaked.

Reply:
Removed and also removed the original in a separate leading patch. This was bad programming by me, I copied that message even though I didn't know why it was there.